client.go

```diff
@@ -86,67 +89,196 @@ func (c *Client) newRequest(ctx context.Context, method, url string, setters ...
+	}
+	return slices.Contains(r.RetryCodes, statusCode)
+}
+
+func (c *Client) sendRequest(req *http.Request, v Response, retryOpts ...RetryOptions) error {
 	req.Header.Set("Accept", "application/json")
 
+	// Default Options
+	options := NewDefaultRetryOptions()
+	options.complete(retryOpts...)
+
 	// Check whether Content-Type is already set, Upload Files API requires
 	// Content-Type == multipart/form-data
 	contentType := req.Header.Get("Content-Type")
 	if contentType == "" {
 		req.Header.Set("Content-Type", "application/json")
 	}
 
-	res, err := c.config.HTTPClient.Do(req)
-	if err != nil {
-		return err
+	const baseDelay = time.Millisecond * 200
+	var (
+		resp     *http.Response
+		err      error
+		failures []string
+	)
+
+	// Save the original request body
+	var bodyBytes []byte
+	if req.Body != nil {
+		bodyBytes, err = io.ReadAll(req.Body)
+		if err != nil {
+			failures = append(failures, fmt.Sprintf("failed to read request body: %v", err))
+			return fmt.Errorf("failed to read request body: %v; failures: %v", err, strings.Join(failures, "; "))
+		}
 	}
 
-	defer res.Body.Close()
+	for i := 0; i <= options.Retries; i++ {
+		// Check if context was canceled (timeout) before retrying
+		if req.Context().Err() != nil {
+			failures = append(failures, fmt.Sprintf("exiting due to canceled context after try #%d/%d: %v", i+1, options.Retries+1, req.Context().Err()))
+			break
+		}
 
-	if isFailureStatusCode(res) {
-		return c.handleErrorResp(res)
-	}
+		// Reset body to the original request body
+		if bodyBytes != nil {
+			req.Body = io.NopCloser(bytes.NewBuffer(bodyBytes))
+		}
 
-	if v != nil {
-		v.SetHeader(res.Header)
-	}
+		resp, err = c.config.HTTPClient.Do(req)
+		if err == nil && !isFailureStatusCode(resp) {
+			defer resp.Body.Close()
+			if v != nil {
+				v.SetHeader(resp.Header)
+			}
+			return decodeResponse(resp.Body, v)
+		}
 
-	return decodeResponse(res.Body, v)
-}
+		// handle connection errors
+		if err != nil {
+			failures = append(failures, fmt.Sprintf("#%d/%d failed to send request: %v", i+1, options.Retries+1, err))
+			continue
+		}
 
-func (c *Client) sendRequestRaw(req *http.Request) (body io.ReadCloser, err error) {
-	resp, err := c.config.HTTPClient.Do(req)
-	if err != nil {
-		return
-	}
+		// handle status codes
+		failures = append(failures, fmt.Sprintf("#%d/%d error response received: %v", i+1, options.Retries+1, c.handleErrorResp(resp)))
 
-	if isFailureStatusCode(resp) {
-		err = c.handleErrorResp(resp)
-		return
+		// exit on non-retriable status codes
+		if !options.canRetry(resp.StatusCode) {
+			failures = append(failures, fmt.Sprintf("exiting due to non-retriable error in try #%d/%d: %v", i+1, options.Retries+1, resp.StatusCode))
+			return fmt.Errorf("request failed on non-retriable error: %v", strings.Join(failures, "; "))
+		}
+
+		// exponential backoff
+		delay := baseDelay * time.Duration(1<<i)
+		jitter := time.Duration(rand.Int63n(int64(baseDelay)))
+		time.Sleep(delay + jitter)
```

Comment:
nit: if you use this, can you get rid of the check at the top of the loop? You'd also have to add an `outer` scope outside the for-loop.
```suggestion
		select {
			case <-ctx.Done():
				failures = append(failures, fmt.Sprintf("exiting due to canceled context after try #%d/%d: %v", i+1, options.Retries+1, req.Context().Err()))
				break outer
			case <-time.Afer(delay + jitter):
		}
```